### PR TITLE
[3D] Add support for a rotated 3d scene

### PR DESCRIPTION
**Pull request by ptitjano**
## Description

It is possible to limit the extent of a 3d view. However, the extents is always axis aligned: it is aligned with the true north. This PR adds the possibility to define a rotation angle around the Z axis. This way, the previously axis aligned extent can now have an arbitrary azimuth. 

Previously, it was possible to fully describe the extent with a `QgsRectangle`. Now, a more general `QgsGeometry` needs to be used. It has the following consequences on `Qgs3DMapSettings`: 

```
There are now 4 stored parameters to fully describe the 3d scene
extent:
- QgsRectangle `mExtent`: axis aligned extent
- double `mZRotation`: Z rotation of the 3d scene
- QgsGeometry `mRotatedExtent`: the rotated `mExtent`
- QgsRectangle `mRotatedExtentBBox`: the bounding box of
  `mRotatedExtent`

If there is no rotation, `mExtent`, `mRotatedExtent` and
`mRotatedExtentBBox` are the same.

`Qgs3DMapSettings::rotatedExtent()` returns a geometry of the rotated
extent.
`Qgs3DMapSettings::extent()` now returns the bounding box of the
rotated extent.
`Qgs3DMapSettings::extent()` now returns the bounding box of the
oriented extent.
```

This works for all the different kind of layers. The change on the 3d loaders is simple. Most of the time, it's enough to update the feature request to take into account the orientation.    
However, the changes on the terrain are much more important. Indeed, the texture of a terrain tile is supposed to be rectangular. This is not the case anymore for a tile on the border and an arbitrary rotation. In order to solve this issue, the idea is to no draw anything for those pixels outside of an rotated extent. This means that those are pixels are transparent. Then, a`QPainterPath` is used to render only the pixels inside the extent. New materials need to be introduced to handle those transparent pixels. They are derived for the usual layers but their fragment shader discard the transparent pixels to not display them and avoid any transparent issue.

Finally, a new 3d extent widget is introduced. It allows to properly set a custom orientation

image

An example: 

image

image

### CODE REVIEW

File: python/3d/auto_generated/qgs3dmapsettings.sip.in [hunk 1]
```diff
@@ -48,12 +48,16 @@ Resolves references to other objects (map layers) after the call to :py:func:`~Q
...
+If the 3D scene is not rotated, this is similar to :py:func:`~Qgs3DMapSettings.orientedExtent`. Otherwise, this
+is the bounding box of :py:func:`~Qgs3DMapSettings.orientedExtent`.
+
+.. seealso:: :py:func:`orientedExtent`
 
 .. versionadded:: 3.30
 %End
 
-    void setExtent( const QgsRectangle &extent );
+    void setExtent( const QgsRectangle &newExtent );
```
Comment (nyalldawson) [inline python/3d/auto_generated/qgs3dmapsettings.sip.in:60]: This is an API break -- it would break existing use of the function with named arguments in Python.

Also, my preference is to the simpler name anyway

File: python/3d/auto_generated/qgs3dmapsettings.sip.in [hunk 2]
```diff
@@ -79,6 +83,29 @@ to improve the precision.
     QgsVector3D origin() const;
 %Docstring
 Returns coordinates in map CRS at which 3D scene has origin (0,0,0)
+%End
+
+    void setOrientation( double newOrientationZ );
```
Comment (nyalldawson) [inline python/3d/auto_generated/qgs3dmapsettings.sip.in:88]:
```suggestion
    void setOrientation( double orientation );
```

Although I'd lean toward setZRotation instead -- we use rotation for this purpose rather than orientation in the API. Orientation is for ring orientation (clockwise / counter clockwise), or page orientation (landscape / portrait)

File: python/3d/auto_generated/qgs3dmapsettings.sip.in [hunk 3]
```diff
@@ -79,6 +83,29 @@ to improve the precision.
...
+%Docstring
+Sets the z orientation of the 3d scene
+
+:param newOrientationZ: orientation in degrees
```
Comment (nyalldawson) [inline python/3d/auto_generated/qgs3dmapsettings.sip.in:92]: Also note if this is clockwise / counterclockwise

File: python/3d/auto_generated/qgs3dmapsettings.sip.in [hunk 4]
```diff
@@ -79,6 +83,29 @@ to improve the precision.
...
+    double orientation() const;
+%Docstring
+Returns the z orientation of the 3d scene
+
+.. versionadded:: 3.38
+%End
+
+    QgsGeometry orientedExtent() const;
+%Docstring
+Returns the 3D scene's 2D geometry in project's CRS.
```
Comment (nyalldawson) [inline python/3d/auto_generated/qgs3dmapsettings.sip.in:106]: Is this correct? Is it actually project crs, even if the 3d map settings object has a different crs() set? Or should this be "in the map's crs()"? (also applies to the other extent methods).

File: src/3d/chunks/qgschunkloader_p.cpp [hunk 5]
```diff
@@ -15,10 +15,50 @@
...
+    {
+      bboxExtent = ct.transformBoundingBox( bboxExtent );
+      sceneExtentGeom.transform( ct );
+    }
+    catch ( QgsCsException & )
+    {
+      QgsDebugError( QStringLiteral( "Error transforming node bounds coordinate" ) );
+    }
+  }
+  QString filterExpression = QStringLiteral( "intersects($geometry, geomFromWkt('%1')) AND intersects($geometry, geomFromWkt('%2'))" ).arg( sceneExtentGeom.asWkt(), bboxExtent.asWktPolygon() );
```
Comment (nyalldawson) [inline src/3d/chunks/qgschunkloader_p.cpp:53]: This approach is a performance killer. It will force a full fetch of the entire layer's features, and then an evaluation on the QGIS client side for the intersection. I.e. it will negate any potential use of the backend's spatial index.

There's two options:

1. Add a geometry filter option to QgsFeatureRequest, and implement for all provider iterators. (Warning: This is a LOT of work.)
2. Use the bounding box of the geometry as a filterRect for the request, and then do the intersection test manually for each fetched feature, using a prepared GEOS geometry of the desired polygon object.

File: src/3d/qgs3dmapsettings.h [hunk 6]
```diff
@@ -98,6 +103,28 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
     //! Returns coordinates in map CRS at which 3D scene has origin (0,0,0)
     QgsVector3D origin() const { return mOrigin; }
 
+    /**
+     * Sets the z orientation of the 3d scene
+     *
+     * \param newOrientationZ orientation in degrees
+     * \since QGIS 3.38
```
Comment (nyalldawson) [inline src/3d/qgs3dmapsettings.h:110]:
```suggestion
     * \since QGIS 3.40
```

elsewhere too

**Comment by nyalldawson** [comment]
@ptitjano this was just a partial review, I thought I'd post some early feedback to get things started :+1:

**Comment by nyalldawson** [comment]
Also @nirvn got any ideas on the new extent widget? I think it'll need a few tweaks to get the same level of UX polish as the existing one.

File: python/3d/auto_generated/qgs3dmapsettings.sip.in (hunk 2 above)
Comment (ptitjano) [reply python/3d/auto_generated/qgs3dmapsettings.sip.in]: `rotation` is now used everywhere.

File: python/3d/auto_generated/qgs3dmapsettings.sip.in (hunk 1 above)
Comment (ptitjano) [reply python/3d/auto_generated/qgs3dmapsettings.sip.in]: Fixed.

File: python/3d/auto_generated/qgs3dmapsettings.sip.in (hunk 3 above)
Comment (ptitjano) [reply python/3d/auto_generated/qgs3dmapsettings.sip.in]: Done.

File: src/3d/qgs3dmapsettings.h (hunk 6 above)
Comment (ptitjano) [reply src/3d/qgs3dmapsettings.h]: Done.

File: python/3d/auto_generated/qgs3dmapsettings.sip.in (hunk 4 above)
Comment (ptitjano) [reply python/3d/auto_generated/qgs3dmapsettings.sip.in:106]: This PR does not change any CRS related thing. I have expanded the existing behavior. Indeed, this is not the case anymore since https://github.com/qgis/qGIS/commit/b0d1a4f8b1cc244ccf2bc3b07798f9ee3540072f

This PR fixes the existing documentation: https://github.com/qgis/QGIS/pull/57613

**Comment by nyalldawson** [comment]
@ptitjano I've had another look over this, and would you mind splitting this into separate modular PRs? Eg I have big concerns about the changes made here to the point cloud request, and would prefer to discuss those in isolation.

File: src/3d/chunks/qgschunkloader_p.cpp (hunk 5 above)
Comment (ptitjano) [reply src/3d/chunks/qgschunkloader_p.cpp]: Thanks. I have done the second option.

**Comment by ptitjano** [comment]
> @ptitjano I've had another look over this, and would you mind splitting this into separate modular PRs? Eg I have big concerns about the changes made here to the point cloud request, and would prefer to discuss those in isolation.

I have created this new PR: https://github.com/qgis/QGIS/pull/57618

**Review by wonder-sk — CHANGES_REQUESTED** [review]
I am afraid this is -1 for me.

This kind of functionality should be really implemented on GPU using clipping planes. OpenGL has a nice builtin functionality for that with `gl_ClipDistance` (and enabled by Qt3DRender::QClipPlane) that even allows one to do clipping in the vertex shader (!), thus saving a lot of processing time because fragment shader does not even need to be called for clipped parts of geometries (e.g. in WebGL this is not available and clipping has to be done in fragment shaders by discarding fragments). This is also a good intro to clipping planes: https://www.youtube.com/watch?v=dcfj5PctptE 

If the clipping is done on GPU, this also means that any change to the clipping configuration is a simple update of some OpenGL uniforms, rather than expensive reload of all entities.

With `gl_ClipDistance`, one would only need small updates in the existing materials to respect clipping planes we would define.

Some more general comments about the current implementation:

- zRotation() - initially I thought all coordinates in the scene were getting rotated by this angle - but in fact the coordinates stay the same (axis aligned with X,Y coordinates of the map projection), just some tricks are applied to clip the scene to make it look like it is rotated

- zRotation() - we (Lutra + North Road) will be hopefully soon adding support to render 3D scene as a globe. In that kind of setting, having a Z-rotation does not really make much sense. If we want a "rotated" scene, we should use some better way to define the clipping area - e.g. a oriented bounding box.

- it seems like unreasonable complication in the chunk loaders to have them all consider the rotated extent (passed as a polygon), especially all the changes in the point cloud infrastructure - as already pointed out by Nyall in #57618. In addition to that, mesh layer does not seem to be covered?

- terrain texture generator + addition of two "masked" materials - this also adds a lot of complexity that would be otherwise not needed with clipping planes approach.

So, to sum up, I think the better and much simpler approach would be to:
- use oriented bounding box to set up clipping for a rotated scene
- leave chunk loading and point cloud infrastructure intact
- update materials to use OpenGL clipping planes

**Comment by ptitjano** [comment]
> I am afraid this is -1 for me.
> 
> This kind of functionality should be really implemented on GPU using clipping planes. OpenGL has a nice builtin functionality for that with `gl_ClipDistance` (and enabled by Qt3DRender::QClipPlane) that even allows one to do clipping in the vertex shader (!), thus saving a lot of processing time because fragment shader does not even need to be called for clipped parts of geometries (e.g. in WebGL this is not available and clipping has to be done in fragment shaders by discarding fragments). This is also a good intro to clipping planes: https://www.youtube.com/watch?v=dcfj5PctptE
> 
> If the clipping is done on GPU, this also means that any change to the clipping configuration is a simple update of some OpenGL uniforms, rather than expensive reload of all entities.
> 
> With `gl_ClipDistance`, one would only need small updates in the existing materials to respect clipping planes we would define.
> 
> Some more general comments about the current implementation:
> 
>     * zRotation() - initially I thought all coordinates in the scene were getting rotated by this angle - but in fact the coordinates stay the same (axis aligned with X,Y coordinates of the map projection), just some tricks are applied to clip the scene to make it look like it is rotated
> 
>     * zRotation() - we (Lutra + North Road) will be hopefully soon adding support to render 3D scene as a globe. In that kind of setting, having a Z-rotation does not really make much sense. If we want a "rotated" scene, we should use some better way to define the clipping area - e.g. a oriented bounding box.
> 
>     * it seems like unreasonable complication in the chunk loaders to have them all consider the rotated extent (passed as a polygon), especially all the changes in the point cloud infrastructure - as already pointed out by Nyall in [3D] Add support for a rotated 3d scene - pointcloud support #57618](https://github.com/qgis/QGIS/pull/57618). In addition to that, mesh layer does not seem to be covered?
> 
>     * terrain texture generator + addition of two "masked" materials - this also adds a lot of complexity that would be otherwise not needed with clipping planes approach.
> 
> 
> So, to sum up, I think the better and much simpler approach would be to:
> 
>     * use oriented bounding box to set up clipping for a rotated scene
> 
>     * leave chunk loading and point cloud infrastructure intact
> 
>     * update materials to use OpenGL clipping planes

I totally agree with you and this now done in https://github.com/qgis/QGIS/pull/57899.    

To be completely honest I wanted to do something similar in the first place. I did not do it because when we (@benoitdm-oslandia and I) first told you about it two years ago, you answered us that this was a bad idea because we would need to change all the shader files. This would be too hard to maintain. Basically, you discouraged us.   I did not agree with the answer but I respected it

I even showed you a first draft of this PR during the QGIS contributor meeting in s-Hertogenbosch. During this meeting, I also explained that it was very hard to add support for the terrain. You never mentioned the shader approach during this meeting. I ended up with the transparent approach which is, I think, a nice hack to solve an issue which should not occur in the first place.

In summary, this is not like this PR is some secret work which comes out of nowhere. I am very disappointed this ends up in a situation like this. This made me waste a lot of time. I hope we can do better in the future. I look forward to it

**Comment by ptitjano** [comment]
The gl_distance approach is now implemented in https://github.com/qgis/QGIS/pull/57899 and https://github.com/qgis/QGIS/pull/58051
